### ISSUE TITLE
Subclassing

### ISSUE THREAD

**Issue by benderTheCrime** [taylorhakes/promise-polyfill#24]
Hey Taylor, long time no see.

What are your thoughts on changing the reference to the Promise constructor in the "then" method instead to `this.constructor`?

This makes it vastly easier to subclass this implementation of Promises because the `Handler` and `handler` functions are enclosed in the closure surrounding the `promise.js` module.

Here's what I did:
```javascript
  Promise.prototype.then = function (onFulfilled, onRejected) {
    var prom = new (this.constructor)(noop);
    handle(this, new Handler(onFulfilled, onRejected, prom));
    return prom;
  };
```
That way, when used in practice with a class:
```
class PromiseTransform extends Promise {
    map() {
        let handler = data => { /* ... */ };
        this.then(handler);
    }
}
```

It allows you to modify the behavior of the then function while still returning a PromiseTransform as opposed to a Promise. `super.then` will return a Promise.

Fortunately, `catch` calls then, so with this change, catch will also return a PromiseTransform as opposed to a Promise. Likewise, `super.catch` will also return a Promise.

I'm fairly confident that if someone wants to do the same with reject or resolve they can handle those subclassed methods themselves.

I ran the tests over this and they all pass.

If you think it's worth doing, I can submit a PR.

**Comment by taylorhakes** [comment]
Hey Joe! That is a great idea. I think we should create a safe then function along the lines of  [`this`](https://github.com/then/promise/blob/master/src/core.js#L73). If you create a pull request, I will definitely accept it. Please add tests as well.

**Comment by benderTheCrime** [comment]
Will do later today, thanks!

**Comment by taylorhakes** [comment]
@benderTheCrime Have you had time to work on this? I think this would be a useful feature. If not, I can work on it.

**Comment by benderTheCrime** [comment]
I'm sorry. I forgot about this altogether. If you have time to work on it that would be great. Otherwise, I can work on it during the week.

**Comment by benderTheCrime** [comment]
Here is a first pass at this: https://github.com/taylorhakes/promise-polyfill/pull/27

I say first pass because I did not see the advantage of defining the `safeThen` once I got down to it. I don't know that all of the work that the `safeThen` method does is necessary to ensure that the Promise that is being resolved is actually a Promise and not an instance of the subclass. Any user subclassing this should be able to use it in such a way that prevents augmentation to any of the private properties created (or at least know when they've broken something sufficiently to backpedal and undo the changes they've made). With this consideration in mind, we prevent ourselves from adding too much code for this functionality.

Also, let me know if the tests added are sufficient. They are pretty basic in comparison to the extent to which the rest of the library is tested.
